### DISCUSSION TITLE
Sometimes, a custom property on socket object is undefined inside disconnect event handler

### DISCUSSION THREAD

**Discussion by awesomelike** [socketio/socket.io#4082] [Q&A] [answered]
I have a custom event `register`, where clients emit on page load. On the server, I get a property `userId` from its payload and set it to the socket object as a property. Because, I will need it in other places. However, __sometimes__ when the client disconnects the customly set property is not there (i.e it is `undefined`). What is wrong with my approach?

```js
class MySocket {
  constructor() {
    io.on('connection', (socket) => {
      socket.on('register', (payload) => this.register(payload, socket));
      socket.on('disconnect', () => this.disconnect(socket));
    });
  }

  register(payload, socket) {
    socket.userId = payload.userId;
  }

  disconnect(socket) {
    console.log(socket.userId); // this is sometimes undefined
  }
}
```

**Comment by bartversluijs** [comment, marked as answer]
Are you sure that `register` is triggered for that socket client before it disconnects?
For debugging purposes, you could log the `socket.id` when the register function is called and check if that `socket.id` has a user ID defined on disconnect. 

I think the `register` event is never triggered, and thus mo user ID is set for that client, before it disconnects.

Another approach would be to save data in a Map, with the socket ID as key and data (like the user ID) as value and removing the key on disconnect. This way, you’ll never be dependent on the socket.io library to handle custom properties.
